FirebaseFunctions/Sources/Callable+Codable.swift

```diff
@@ -160,3 +165,183 @@ public struct Callable<Request: Encodable, Response: Decodable> {
     return try await call(data)
   }
 }
+
+/// Used to determine when a `StreamResponse<_, _>` is being decoded.
+private protocol StreamResponseProtocol {}
+
+/// A convenience type used to receive both the streaming callable function's
+/// yielded messages and its return value.
+///
+/// This can be used as the generic `Response` parameter to ``Callable`` to
+/// receive both the yielded messages and final return value of the streaming
+/// callable function.
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+public enum StreamResponse<Message: Decodable, Result: Decodable>: Decodable,
+  StreamResponseProtocol {
+  /// The message yielded by the callable function.
+  case message(Message)
+  /// The final result returned by the callable function.
+  case result(Result)
+
+  private enum CodingKeys: String, CodingKey {
+    case message
+    case result
+  }
+
+  public init(from decoder: any Decoder) throws {
+    do {
+      let container = try decoder
+        .container(keyedBy: Self<Message, Result>.CodingKeys.self)
+      var allKeys = ArraySlice(container.allKeys)
+      guard let onlyKey = allKeys.popFirst(), allKeys.isEmpty else {
```

Comment:
nit: I think checking for `count == 1` explicitly might be simpler than the `ArraySlice` approach.
```suggestion
      guard container.allKeys.count == 1, let onlyKey = container.allKeys.first else {
```